auth.php

```diff
@@ -69,6 +71,50 @@ public function __construct() {
     }
 
     /**
+     * All the checking happens before the login page in this hook
+     */
+    public function pre_loginpage_hook() {
+        global $SESSION;
+
+        // If we previously tried to skip SSO on, but then navigated
+        // away, and come in from another deep link while SSO only is
+        // on, then reset the previous session memory of forcing SSO.
+        if (isset($SESSION->skipsso)) {
+            unset($SESSION->skipsso);
+        }
+
+        $this->loginpage_hook();
+    }
+
+    /**
+     * All the checking happens before the login page in this hook
+     */
+    public function loginpage_hook() {
+        global $SESSION;
+        $skipsso = optional_param('skipsso', 0, PARAM_BOOL);
+
+        // Check whether we've skipped SSO already.
+        // This is here because loginpage_hook is called again during form
+        // submission (all of login.php is processed) and ?skipsso=on is not
+        // preserved forcing us to the SSO.
+        if ((isset($SESSION->skipsso) && $SESSION->skipsso == 1)) {
+            return;
+        }
+
+        $SESSION->skipsso = $skipsso;
+
+        // If SSO only is set and user is not passing the skip param
+        // or has it already set in their session then redirect to the SSO URL.
+        if ($this->config->ssoonly == 1 && !$skipsso) {
+            redirect($this->config->ssourl);
```

Comment:
What's going to happen if ssourl is empty? We need to add this check when we save config. If we set ssoonly = yes, then we have to add ssourl.